modules/core/validator.js

```diff
@@ -176,24 +176,23 @@ export function coreValidator(context) {
   validator.getIssues = (options) => {
     const opts = Object.assign({ what: 'all', where: 'all', includeIgnored: false, includeDisabledRules: false }, options);
     const view = context.map().extent();
+    const baseGraph = context.history().base();
     let issues = [];
     let seen = new Set();
 
     // collect head issues - caused by user edits
-    let cache = _headCache;
-    if (_headGraph) {
-      Object.values(cache.issuesByIssueID).forEach(issue => {
-        if (!filter(issue, _headGraph, cache)) return;
+    if (_headGraph && _headGraph !== baseGraph) {
```

Comment:
I have a couple of questions on this:
1) In general, If some entity with already existing issue was edited, should this issue be treated as a new/'head' issue and be previewed in warnings tab and on commit page? I've observed that iD v2.19.6's answer to this was 'no'.
2) If we go with 'yes', then we should be carefull in cases such as these. I don't think that the current logic handles the following:
0. Position near an entity (e1) that has an already existing issue + open warnings tab (tab: no issues)
1. Make a change on an entity that does not resolve an issue (tab: e1's issues)
2. Undo (tab: no issues)
3. Make any other change (add new entity etc.) (tab: e1's issues) - [expected (no issues)]
In this case most likely additional logic would have to be added that would unload/load head_cache on undo/redo.

Reply:
> In general, If some entity with already existing issue was edited, should this issue be treated as a new/'head' issue and be previewed in warnings tab and on commit page? I've observed that iD v2.19.6's answer to this was 'no'.

I think yes, if a user edits a feature and creates and issue with it, it should count as a new issue.  iD 2.19.6 did have trouble counting issues accurately because iirc it just ran all the validations twice on every edit.

> If we go with 'yes', then we should be carefull in cases such as these. I don't think that the current logic handles the following:
> 0. Position near an entity (e1) that has an already existing issue + open warnings tab (tab: no issues)
> 1. Make a change on an entity that does not resolve an issue (tab: e1's issues)
> 2. Undo (tab: no issues)
> 3. Make any other change (add new entity etc.) (tab: e1's issues) - [expected (no issues)]
> 4. In this case most likely additional logic would have to be added that would unload/load head_cache on undo/redo.

I think `validateEntitiesAsync()` is supposed to uncache the detected issues before validation, but I do see that across undo it doesn't completely remove them all.  I'll investigate and try to fix, but would prefer for this not to be a blocker to these other improvements.

Reply:
Ok, I won't block this PR as there are several good changes done.
It is possible you didn't fully catch the case I am describing, so I will create a separate ticket and link it here to track this.

Reply:
https://github.com/openstreetmap/iD/issues/8632